osfoffline/utils/authentication.py

```diff
@@ -0,0 +1,120 @@
+import asyncio
+import datetime
+import json
+import logging
+
+import aiohttp
+import furl
+from sqlalchemy.exc import SQLAlchemyError
+
+from osfoffline import settings
+from osfoffline.database_manager.db import session
+from osfoffline.database_manager.models import User
+from osfoffline.database_manager.utils import save
+from osfoffline.polling_osf_manager.remote_objects import RemoteUser
+from osfoffline.exceptions import AuthError
+
+class AuthClient(object):
+    """Manages authorization flow """
+
+    @asyncio.coroutine
+    def _authenticate(self, username, password):
+        """ Tries to use standard auth to authenticate and create a personal access token through APIv2
+
+            :return: personal_access_token or raise AuthError
+        """
+        token_url = furl.furl(settings.API_BASE)
+        token_url.path.add('/v2/tokens/')
+        token_request_body = {
+            'data': {
+                'type': 'tokens',
+                'attributes': {
+                    'name': 'OSF-Offline - {}'.format(datetime.date.today()),
+                    'scopes': settings.APPLICATION_SCOPES
+                }
+            }
+        }
+        headers = {'content-type': 'application/json'}
+
+        try:
+            resp = yield from aiohttp.request(method='POST', url=token_url.url, headers=headers, data=json.dumps(token_request_body), auth=(username, password))
```

Comment:
This is a **very** low priority bug right now, but this login method will not allow users log in if their password contains certain special characters (like the unicode snowman) that are considered valid password characters on the OSF.

It fails immediately before contacting the server with the traceback below. The root cause is that aiohttp's basic auth header [defaults](http://aiohttp.readthedocs.org/en/stable/api.html#aiohttp.helpers.BasicAuth) to using latin1 encoding. 

```
[2015-11-14 12:55:35,332][MainThread][start_screen.py][ERROR][root]: Login failed
Traceback (most recent call last):
  File "/Users/andyboughton/code/cos/OSF-Offline/osfoffline/utils/authentication.py", line 40, in _authenticate
    resp = yield from aiohttp.request(method='POST', url=token_url.url, headers=headers, data=json.dumps(token_request_body), auth=(username, password))
  File "/Users/andyboughton/code/cos/OSF-Offline/env/lib/python3.5/site-packages/aiohttp/client.py", line 456, in __iter__
    resp = yield from self._coro
  File "/Users/andyboughton/code/cos/OSF-Offline/env/lib/python3.5/site-packages/aiohttp/client.py", line 171, in _request
    loop=self._loop, response_class=self._response_class)
  File "/Users/andyboughton/code/cos/OSF-Offline/env/lib/python3.5/site-packages/aiohttp/client_reqrep.py", line 88, in __init__
    self.update_auth(auth)
  File "/Users/andyboughton/code/cos/OSF-Offline/env/lib/python3.5/site-packages/aiohttp/client_reqrep.py", line 254, in update_auth
    self.headers[hdrs.AUTHORIZATION] = auth.encode()
  File "/Users/andyboughton/code/cos/OSF-Offline/env/lib/python3.5/site-packages/aiohttp/helpers.py", line 34, in encode
    creds = ('%s:%s' % (self.login, self.password)).encode(self.encoding)
UnicodeEncodeError: 'latin-1' codec can't encode character '\u2603' in position 20: ordinal not in range(256)
```
The effort required to fix it would be disproportionate relative to the likely payoff, but worth knowing about in case support gets asked.